src/pal/src/misc/miscpalapi.cpp

```diff
@@ -226,86 +230,154 @@ PAL_GetPALDirectoryA(
     return bRet;
 }
 
-VOID
-PALAPI
-PAL_Random(
+static
+BOOL
+PAL_RandomFromUrandom(
         IN OUT LPVOID lpBuffer,
         IN DWORD dwLength)
 {
-    int rand_des = -1;
-    DWORD i;
-    long num = 0;
     static BOOL sMissingDevURandom;
-    static BOOL sInitializedMRand;
+    int rand_des;
 
-    PERF_ENTRY(PAL_Random);
-    ENTRY("PAL_Random(lpBuffer=%p, dwLength=%d)\n", lpBuffer, dwLength);
+    if (sMissingDevURandom)
+    {
+        return FALSE;
+    }
+
+    do
+    {
+        rand_des = open(URANDOM_DEVICE_NAME, O_RDONLY | O_CLOEXEC);
+    } while ((rand_des == -1) && (errno == EINTR));
 
-    if (!sMissingDevURandom)
+    if (rand_des == -1)
     {
-        do
+        if (errno == ENOENT)
         {
-            rand_des = open("/dev/urandom", O_RDONLY, O_CLOEXEC);
+            sMissingDevURandom = TRUE;
         }
-        while ((rand_des == -1) && (errno == EINTR));
+        else
+        {
+            ASSERT("PAL__open() failed, errno:%d (%s)\n", errno, strerror(errno));
+        }
+
+        return FALSE;
+    }
 
-        if (rand_des == -1)
+    DWORD offset = 0;
+    do
+    {
+        ssize_t n = read(rand_des, (BYTE*)lpBuffer + offset , dwLength - offset);
+        if (n == -1)
         {
-            if (errno == ENOENT)
+            if (errno == EINTR)
             {
-                sMissingDevURandom = TRUE;
-            }
-            else
-            {
-                ASSERT("PAL__open() failed, errno:%d (%s)\n", errno, strerror(errno));
+                continue;
             }
+            ASSERT("read() failed, errno:%d (%s)\n", errno, strerror(errno));
 
-            // Back off and try mrand48.
+            break;
         }
-        else
+
+        offset += n;
+    }
+    while (offset != dwLength);
+
+    _ASSERTE(offset == dwLength);
+
+    close(rand_des);
+
+    return TRUE;
+}
+
+#if defined(HAVE_GETRANDOM)
+static
+BOOL
+PAL_RandomFromGetRandom(
+        IN OUT LPVOID lpBuffer,
+        IN DWORD dwLength)
+{
+    ssize_t length = (ssize_t)dwLength;
+    static BOOL sMissingGetRandom;
+
+    if (sMissingGetRandom)
+    {
+        return FALSE;
+    }
+
+    do
+    {
+        ssize_t res = getrandom(lpBuffer, dwLength, GRND_NONBLOCK);
+        if (res == -1)
         {
-            DWORD offset = 0;
-            do
+            if (errno == ENOSYS)
             {
-                ssize_t n = read(rand_des, (BYTE*)lpBuffer + offset , dwLength - offset);
-                if (n == -1)
-                {
-                    if (errno == EINTR)
-                    {
-                        continue;
-                    }
-                    ASSERT("read() failed, errno:%d (%s)\n", errno, strerror(errno));
-
-                    break;
-                }
-
-                offset += n;
+                sMissingGetRandom = TRUE;
+                return FALSE;
+            }
+            else if ((errno == EINTR) || (errno == EAGAIN))
+            {
+                continue;
```

Comment:
Do you know what the getrandom doc means by "if the urandom source has not yet been initialized"? Does it mean that enough entropy was not accumulated or is it something else? I've seen notes that waiting for the entropy pool initialization may take non-trivial amount of time. If that applied here, I would just fail in case it would block.
The PAL_Random is not supposed to be used for any cryptographical stuff, we use it to generate Guids for internal purposes (debugger) and random numbers in superpmi jit tests. Blocking would not be desired here.

Reply:
This is often just in early boot situations where there are no hardware-provided entropy sources. For systems with RDRAND (or equivalent), networked devices (so lots of interrupt activity going on), etc. this isn't much of an issue.  In any case, I've modified the code so it doesn't retry on EAGAIN, only on EINTR -- so it's closer to what the /dev/urandom function does.